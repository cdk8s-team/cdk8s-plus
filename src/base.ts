import { ApiObjectMetadata, ApiObject, ApiObjectMetadataDefinition } from 'cdk8s';
import { Construct } from 'constructs';
import { IApiResource, IApiEndpoint } from './api-resource.generated';
import * as r from './role';
import * as rb from './role-binding';

/**
 * Initialization properties for resources.
 */
export interface ResourceProps {
  /**
   * Metadata that all persisted resources must have, which includes all objects
   * users must create.
   */
  readonly metadata?: ApiObjectMetadata;
}

/**
 * Represents a resource.
 */
export interface IResource {
  /**
   * The Kubernetes name of this resource.
   */
  readonly name: string;

  /**
   * The object's API version (e.g. "authorization.k8s.io/v1")
   */
  readonly apiVersion: string;

  /**
   * The group portion of the API version (e.g. "authorization.k8s.io")
   */
  readonly apiGroup: string;

  /**
   * The object kind (e.g. "Deployment").
   */
  readonly kind: string;
}

/**
 * Base class for all Kubernetes objects in stdk8s. Represents a single
 * resource.
 */
export abstract class Resource extends Construct implements IResource, IApiResource, IApiEndpoint {

  /**
   * The underlying cdk8s API object.
   */
  protected abstract readonly apiObject: ApiObject;

  public get metadata(): ApiObjectMetadataDefinition {
    return this.apiObject.metadata;
  }

  /**
   * The name of this API object.
   */
  public get name(): string {
    return this.apiObject.name;
  }

  /**
   * The object's API version (e.g. "authorization.k8s.io/v1")
   */
  public get apiVersion(): string {
    return this.apiObject.apiVersion;
  }

  /**
   * The group portion of the API version (e.g. "authorization.k8s.io").
   */
  public get apiGroup(): string {
    return this.apiObject.apiGroup;
  }

  /**
   * The object kind (e.g. "Deployment").
   */
  public get kind(): string {
    return this.apiObject.kind;
  }

  public get resourceName(): string | undefined {
    return this.name;
  }

  public readonly abstract resourceType: string;

  public asApiResource(): IApiResource | undefined {
    return this;
  }

  public asNonApiResource(): string | undefined {
    return undefined;
  }

  /**
   * Grants the list of subjects permissions to read this resource.
   */
  public grantRead(...subjects: rb.ISubject[]): rb.RoleBinding {
    const role = new r.Role(this, 'Role', {
      metadata: { namespace: this.metadata.namespace },
    });
    role.allowRead(this);
    return role.bind(...subjects);
  }

  /**
   * Grants the list of subjects permissions to read and write this resource.
   */
  public grantReadWrite(...subjects: rb.ISubject[]): rb.RoleBinding {
    const role = new r.Role(this, 'Role', {
      metadata: { namespace: this.metadata.namespace },
    });
    role.allowReadWrite(this);
    return role.bind(...subjects);
  }

}