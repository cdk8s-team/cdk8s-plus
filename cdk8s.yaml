output: dist
pluginsDirectory: /home/runner/.cdk8s/plugins
imports:
  - k8s@1.27.0
  - k8s@1.28.0
