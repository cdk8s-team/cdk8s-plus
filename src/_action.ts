import { Container } from './container';
import * as k8s from './imports/k8s';

/**
 * Utility class to implement the conversion between our API and the k8s action
 * structure. Used both for probes and handlers.
 *
 * @internal
 */
export class Action {

  public static fromTcpSocket(container: Container, options: { port?: number; host?: string } = {}): k8s.TcpSocketAction {
    return {
      port: k8s.IntOrString.fromNumber(options.port ?? container.port ?? 80),
      host: options.host,
    };
  }

  public static fromCommand(command: string[]): k8s.ExecAction {
    return { command };
  }

  public static fromHttpGet(container: Container, path: string, options: { port?: number } = { }): k8s.HttpGetAction {
    return {
      path,
      port: k8s.IntOrString.fromNumber(options.port ?? container.port ?? 80),
    };
  }
}