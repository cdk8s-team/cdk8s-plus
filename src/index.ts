export * from './base';
export * from './config-map';
export * from './container';
export * from './deployment';
export * from './job';
export * from './pod';
export * from './secret';
export * from './service-account';
export * from './service';
export * from './stateful-set';
export * from './volume';
export * from './ingress';
export * from './probe';
export * from './pvc';
export * from './pv';
export * from './handler';
export * from './workload';
export * from './daemon-set';
export * from './role';
export * from './role-binding';
export * from './namespace';

export * from './api-resource.generated';
