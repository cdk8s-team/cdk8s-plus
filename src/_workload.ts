import { ApiObjectMetadata, ApiObjectMetadataDefinition } from 'cdk8s';
import { Construct } from 'constructs';
import { Resource, ResourceProps } from './base';
import { Container, ContainerProps } from './container';
import { HostAlias, PodSecurityContext, PodSpec, PodSpecProps, RestartPolicy } from './pod';
import { IServiceAccount } from './service-account';
import { Volume } from './volume';

/**
 * Properties for `Workload`.
 */
export interface WorkloadProps extends ResourceProps, PodSpecProps {

  /**
   * The pod metadata of this workload.
   */
  readonly podMetadata?: ApiObjectMetadata;

}

export abstract class Workload extends Resource {

  protected readonly podSpec: PodSpec;

  /**
   * The metadata of pods in this deployment.
   */
  public readonly podMetadata: ApiObjectMetadataDefinition;

  constructor(scope: Construct, id: string, props: WorkloadProps = {}) {
    super(scope, id);

    this.podSpec = new PodSpec(props);
    this.podMetadata = new ApiObjectMetadataDefinition(props.podMetadata);
  }

  public get containers(): Container[] {
    return this.podSpec.containers;
  }

  public get initContainers(): Container[] {
    return this.podSpec.initContainers;
  }

  public get volumes(): Volume[] {
    return this.podSpec.volumes;
  }

  public get restartPolicy(): RestartPolicy | undefined {
    return this.podSpec.restartPolicy;
  }

  public get serviceAccount(): IServiceAccount | undefined {
    return this.podSpec.serviceAccount;
  }

  public get securityContext(): PodSecurityContext {
    return this.podSpec.securityContext;
  }

  public get hostAliases(): HostAlias[] {
    return this.podSpec.hostAliases;
  }

  public addContainer(container: ContainerProps): Container {
    return this.podSpec.addContainer(container);
  }

  public addInitContainer(container: ContainerProps): Container {
    return this.podSpec.addInitContainer(container);
  }

  public addVolume(volume: Volume): void {
    return this.podSpec.addVolume(volume);
  }

  public addHostAlias(hostAlias: HostAlias): void {
    return this.podSpec.addHostAlias(hostAlias);
  }

}