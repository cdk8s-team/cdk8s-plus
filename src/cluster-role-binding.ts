import { ApiObject, Lazy } from 'cdk8s';
import { Construct } from 'constructs';
import { Resource, ResourceProps } from './base';
import * as clusterrole from './cluster-role';
import * as k8s from './imports/k8s';
import * as rb from './role-binding';
import { filterUndefined } from './utils';

/**
 * Properties for `ClusterRoleBinding`.
 */
export interface ClusterRoleBindingProps extends ResourceProps {
  /**
   * The role to bind to.
   */
  readonly role: clusterrole.IClusterRole;
}

/**
 * A ClusterRoleBinding grants permissions cluster-wide to a user or
 * set of users.
 */
export class ClusterRoleBinding extends Resource {
  /**
   * @see base.Resource.apiObject
   */
  protected readonly apiObject: ApiObject;

  public readonly resourceType = 'clusterrolebindings';

  public readonly role: clusterrole.IClusterRole;

  private readonly _subjects: Array<rb.ISubject>;

  constructor(scope: Construct, id: string, props: ClusterRoleBindingProps) {
    super(scope, id);

    this.role = props.role;

    this._subjects = new Array<rb.ISubject>();

    this.apiObject = new k8s.KubeClusterRoleBinding(this, 'Resource', {
      metadata: props.metadata,
      subjects: Lazy.any({ produce: () => this.synthesizeSubjects() }),
      roleRef: Lazy.any({ produce: () => this.synthesizeRoleRef() }),
    });
  }

  /**
   * Adds a subject to the role.
   * @param subjects The subjects to add
   */
  public addSubjects(...subjects: rb.ISubject[]): void {
    for (const subject of subjects) {
      this._subjects.push(subject);
    }
  }

  public get subjects() {
    return [...this._subjects];
  }

  private synthesizeRoleRef(): k8s.RoleRef {
    return {
      apiGroup: this.role.apiGroup,
      kind: this.role.kind,
      name: this.role.name,
    };
  }

  private synthesizeSubjects(): k8s.Subject[] {
    return this._subjects.map((subject) => filterUndefined({
      apiGroup: subject.apiGroup === 'core' ? '' : subject.apiGroup,
      kind: subject.kind,
      name: subject.name,
      namespace: subject.namespace,
    }));
  }
}
