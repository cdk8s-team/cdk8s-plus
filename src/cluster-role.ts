import { ApiObject, Lazy, Names } from 'cdk8s';
import { Construct } from 'constructs';
import { IApiEndpoint } from './api-resource.generated';
import * as base from './base';
import * as clusterrb from './cluster-role-binding';
import * as k8s from './imports/k8s';
import * as role from './role';
import * as rb from './role-binding';

/**
 * Represents a cluster-level role.
 */
export interface IClusterRole extends base.IResource {}

/**
 * Policy rule of a `ClusterRole.
 */
export interface ClusterRolePolicyRule {

  /**
   * Verbs to allow. (e.g ['get', 'watch'])
   */
  readonly verbs: string[];

  /**
   * Endpoints this rule applies to. Can be either api resources
   * or non api resources.
   */
  readonly endpoints: IApiEndpoint[];
}

/**
 * Properties for `ClusterRole`.
 */
export interface ClusterRoleProps extends base.ResourceProps {

  /**
   * A list of rules the role should allow.
   *
   * @default []
   */
  readonly rules?: ClusterRolePolicyRule[];

  /**
   * Specify labels that should be used to locate ClusterRoles, whose rules
   * will be automatically filled into this ClusterRole's rules.
   */
  readonly aggregationLabels?: { [key: string]: string };
}

/**
 * ClusterRole is a cluster level, logical grouping of PolicyRules that can be
 * referenced as a unit by a RoleBinding or ClusterRoleBinding.
 */
export class ClusterRole extends base.Resource implements IClusterRole, role.IRole {

  /**
   * Imports a role from the cluster as a reference.
   * @param name The name of the role resource.
   */
  public static fromClusterRoleName(name: string): IClusterRole {
    return {
      apiGroup: 'rbac.authorization.k8s.io',
      name,
      ...k8s.KubeClusterRole.GVK,
    };
  }

  /**
   * @see base.Resource.apiObject
   */
  protected readonly apiObject: ApiObject;

  public readonly resourceType = 'clusterroles';

  private readonly _labelSelector: Record<string, string> = {};
  private readonly _rules: Array<k8s.PolicyRule> = [];

  constructor(scope: Construct, id: string, props: ClusterRoleProps = {}) {
    super(scope, id);

    this.apiObject = new k8s.KubeClusterRole(this, 'Resource', {
      metadata: props.metadata,
      rules: Lazy.any({ produce: () => this._rules }),
      aggregationRule: Lazy.any({ produce: () => this.synthesizeAggregationRules() }),
    });

    for (const rule of props.rules ?? []) {
      this.allow(rule.verbs, ...rule.endpoints);
    }

    for (const [key, value] of Object.entries(props.aggregationLabels ?? {})) {
      this.aggregate(key, value);
    }
  }

  /**
   * Add permission to perform a list of HTTP verbs on a collection of
   * resources.
   *
   * @param endpoints The endpoints(s) to apply to
   * @see https://kubernetes.io/docs/reference/access-authn-authz/authorization/#determine-the-request-verb
   */
  public allow(verbs: string[], ...endpoints: IApiEndpoint[]): void {
    for (const endpoint of endpoints) {

      const resource = endpoint.asApiResource();
      const nonResource = endpoint.asNonApiResource();

      if (resource && nonResource) {
        throw new Error('Endpoint must be either resource or non resource. not both.');
      }

      if (!resource && !nonResource) {
        throw new Error('Endpoint must be either resource or non resource. not neither.');
      }

      if (resource) {
        this._rules.push({
          apiGroups: [resource.apiGroup === 'core' ? '' : resource.apiGroup],
          resources: [resource.resourceType],
          resourceNames: resource.resourceName ? [resource.resourceName] : [],
          verbs,
        });
      }
      if (nonResource) {
        this._rules.push({ verbs, nonResourceUrLs: [nonResource] });
      }
    }
  }

  /**
   * Add "create" permission for the resources.
   * @param endpoints The resource(s) to apply to
   */
  public allowCreate(...endpoints: IApiEndpoint[]): void {
    this.allow(['create'], ...endpoints);
  }

  /**
   * Add "get" permission for the resources.
   * @param endpoints The resource(s) to apply to
   */
  public allowGet(...endpoints: IApiEndpoint[]): void {
    this.allow(['get'], ...endpoints);
  }

  /**
   * Add "list" permission for the resources.
   * @param endpoints The resource(s) to apply to
   */
  public allowList(...endpoints: IApiEndpoint[]): void {
    this.allow(['list'], ...endpoints);
  }

  /**
   * Add "watch" permission for the resources.
   * @param endpoints The resource(s) to apply to
   */
  public allowWatch(...endpoints: IApiEndpoint[]): void {
    this.allow(['watch'], ...endpoints);
  }

  /**
   * Add "update" permission for the resources.
   * @param endpoints The resource(s) to apply to
   */
  public allowUpdate(...endpoints: IApiEndpoint[]): void {
    this.allow(['update'], ...endpoints);
  }

  /**
   * Add "patch" permission for the resources.
   * @param endpoints The resource(s) to apply to
   */
  public allowPatch(...endpoints: IApiEndpoint[]): void {
    this.allow(['patch'], ...endpoints);
  }

  /**
   * Add "delete" permission for the resources.
   * @param endpoints The resource(s) to apply to
   */
  public allowDelete(...endpoints: IApiEndpoint[]): void {
    this.allow(['delete'], ...endpoints);
  }

  /**
   * Add "deletecollection" permission for the resources.
   * @param endpoints The resource(s) to apply to
   */
  public allowDeleteCollection(...endpoints: IApiEndpoint[]): void {
    this.allow(['deletecollection'], ...endpoints);
  }

  /**
   * Add "get", "list", and "watch" permissions for the resources.
   * @param endpoints The resource(s) to apply to
   */
  public allowRead(...endpoints: IApiEndpoint[]): void {
    this.allow(['get', 'list', 'watch'], ...endpoints);
  }

  /**
   * Add "get", "list", "watch", "create", "update", "patch", "delete", and
   * "deletecollection" permissions for the resources.
   *
   * @param endpoints The resource(s) to apply to
   */
  public allowReadWrite(...endpoints: IApiEndpoint[]): void {
    this.allow(['get', 'list', 'watch', 'create', 'update', 'patch', 'delete', 'deletecollection'], ...endpoints);
  }

  /**
   * Aggregate rules from roles matching this label selector.
   */
  public aggregate(key: string, value: string): void {
    this._labelSelector[key] = value;
  }

  /**
   * Combines the rules of the argument ClusterRole into this ClusterRole
   * using aggregation labels.
   * @param rol
   */
  public combine(rol: ClusterRole): void {
    const key = `cdk8s.cluster-role/aggregate-to-${Names.toLabelValue(this)}`;
    const value = 'true';
    rol.metadata.addLabel(key, value);
    this.aggregate(key, value);
  }

  /**
   * Create a RoleBinding that binds the permissions in this ClusterRole
   * to a list of subjects, that will only apply to the given namespace.
   * @param namespace the namespace to limit permissions to.
   * @param subjects a list of subjects to bind to
   */
  public bindInNamespace(namespace: string, ...subjects: rb.ISubject[]): rb.RoleBinding {
    const binding = new rb.RoleBinding(this, `RoleBinding-${namespace}`, {
      metadata: {
        namespace,
      },
      role: this,
    });
    binding.addSubjects(...subjects);
    return binding;
  }

  /**
   * Create a ClusterRoleBinding that binds the permissions in this
   * ClusterRole to a list of subjects, without namespace restrictions.
   * @param subjects a list of subjects to bind to
   */
  public bind(...subjects: rb.ISubject[]): clusterrb.ClusterRoleBinding {
    const binding = new clusterrb.ClusterRoleBinding(this, 'ClusterRoleBinding', {
      role: this,
    });
    binding.addSubjects(...subjects);
    return binding;
  }

  private synthesizeAggregationRules(): k8s.AggregationRule | undefined {
    if (Object.keys(this._labelSelector).length === 0) {
      return undefined;
    }

    return { clusterRoleSelectors: [{ matchLabels: this._labelSelector }] };
  }
}