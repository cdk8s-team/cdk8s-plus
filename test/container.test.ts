import * as cdk8s from 'cdk8s';
import { Size, Testing } from 'cdk8s';
import * as kplus from '../src';
import { Container, Cpu, Handler } from '../src';
import * as k8s from '../src/imports/k8s';

describe('EnvValue', () => {

  test('Can be created from value', () => {

    const actual = kplus.EnvValue.fromValue('value');

    expect(actual.value).toEqual('value');
    expect(actual.valueFrom).toBeUndefined();

  });

  test('Can be created from config map name', () => {

    const actual = kplus.EnvValue.fromConfigMap(kplus.ConfigMap.fromConfigMapName('ConfigMap'), 'key');

    expect(actual.value).toBeUndefined();
    expect(actual.valueFrom).toEqual({
      configMapKeyRef: {
        key: 'key',
        name: 'ConfigMap',
      },
    });

  });

  test('Can be created from secret value', () => {
    const secretValue = {
      secret: kplus.Secret.fromSecretName('Secret'),
      key: 'my-key',
    };

    const actual = kplus.EnvValue.fromSecretValue(secretValue);

    expect(actual.value).toBeUndefined();
    expect(actual.valueFrom).toEqual({
      secretKeyRef: {
        key: 'my-key',
        name: 'Secret',
      },
    });
  });

  test('Cannot be created from missing required process env', () => {

    const key = 'cdk8s-plus.tests.container.env.fromProcess';
    expect(() => kplus.EnvValue.fromProcess(key, { required: true })).toThrowError(`Missing ${key} env variable`);

  });

  test('Can be created from missing optional process env', () => {

    const key = 'cdk8s-plus.tests.container.env.fromProcess';
    const actual = kplus.EnvValue.fromProcess(key);

    expect(actual.value).toBeUndefined();
    expect(actual.valueFrom).toBeUndefined();

  });

  test('Can be created from existing process env', () => {

    const key = 'cdk8s-plus.tests.container.env.fromProcess';
    try {
      process.env[key] = 'value';
      const actual = kplus.EnvValue.fromProcess(key);

      expect(actual.value).toEqual('value');
      expect(actual.valueFrom).toBeUndefined();

    } finally {
      delete process.env[key];
    }

  });

  test('Can be created from fieldRef', () => {
    const actual = kplus.EnvValue.fromFieldRef(kplus.EnvFieldPaths.POD_NAME);

    expect(actual.value).toBeUndefined();
    expect(actual.valueFrom).toEqual({
      fieldRef: {
        fieldPath: 'metadata.name',
      },
    });
  });

  test('Can be created from fieldRef with key', () => {
    const actual = kplus.EnvValue.fromFieldRef(kplus.EnvFieldPaths.POD_LABEL, { key: 'someLabel' });

    expect(actual.value).toBeUndefined();
    expect(actual.valueFrom).toEqual({
      fieldRef: {
        fieldPath: 'metadata.labels[\'someLabel\']',
      },
    });
  });

  test('Can not be created from fieldRef without key', () => {
    expect(() => kplus.EnvValue.fromFieldRef(kplus.EnvFieldPaths.POD_LABEL)).toThrowError(`${kplus.EnvFieldPaths.POD_LABEL} requires a key`);
  });

  test('Can be created from resourceFieldRef', () => {
    const actual = kplus.EnvValue.fromResource(kplus.ResourceFieldPaths.CPU_LIMIT);

    expect(actual.value).toBeUndefined();
    expect(actual.valueFrom).toEqual({
      resourceFieldRef: {
        resource: 'limits.cpu',
      },
    });
  });

  test('Can be created from resourceFieldRef with divisor', () => {
    const actual = kplus.EnvValue.fromResource(kplus.ResourceFieldPaths.MEMORY_LIMIT, { divisor: '1Mi' });

    expect(actual.value).toBeUndefined();
    expect(actual.valueFrom).toEqual({
      resourceFieldRef: {
        resource: 'limits.memory',
        divisor: k8s.IntOrString.fromString('1Mi'),
      },
    });
  });

  test('Can be created from resourceFieldRef with container', () => {
    const container = new kplus.Container({
      image: 'image',
      name: 'name',
      imagePullPolicy: kplus.ImagePullPolicy.NEVER,
    });

    const actual = kplus.EnvValue.fromResource(kplus.ResourceFieldPaths.CPU_LIMIT, { container });

    expect(actual.value).toBeUndefined();
    expect(actual.valueFrom).toEqual({
      resourceFieldRef: {
        resource: 'limits.cpu',
        containerName: container.name,
      },
    });
  });
});

describe('Container', () => {

  test('Instantiation properties are all respected', () => {

    const container = new kplus.Container({
      image: 'image',
      name: 'name',
      imagePullPolicy: kplus.ImagePullPolicy.NEVER,
      workingDir: 'workingDir',
      port: 9000,
      command: ['command'],
      envVariables: {
        key: kplus.EnvValue.fromValue('value'),
      },
    });

    const actual: k8s.Container = container._toKube();

    const expected: k8s.Container = {
      name: 'name',
      imagePullPolicy: 'Never',
      image: 'image',
      workingDir: 'workingDir',
      ports: [{
        containerPort: 9000,
      }],
      command: ['command'],
      env: [{
        name: 'key',
        value: 'value',
      }],
      securityContext: {
        privileged: false,
        readOnlyRootFilesystem: false,
        runAsNonRoot: false,
      },
    };

    expect(actual).toEqual(expected);

  });

  test('Must use container props', () => {

    const container = new kplus.Container({
      image: 'image',
    });
    expect(() => {
      new kplus.Container(container);
    }).toThrow();

  });

  test('Can add environment variable', () => {

    const container = new kplus.Container({
      image: 'image',
    });

    container.env.addVariable('key', kplus.EnvValue.fromValue('value'));

    const actual: k8s.EnvVar[] = container._toKube().env!;
    const expected: k8s.EnvVar[] = [{
      name: 'key',
      value: 'value',
      valueFrom: undefined,
    }];

    expect(actual).toEqual(expected);

  });

  test('can add environment variables from a source', () => {

    const chart = Testing.chart();

    const cm = new kplus.ConfigMap(chart, 'ConfigMap');
    const secret = new kplus.Secret(chart, 'Secret');

    const cmSource = kplus.Env.fromConfigMap(cm, 'pref');
    const secretSource = kplus.Env.fromSecret(secret);

    const container = new kplus.Container({
      image: 'image',
      envFrom: [cmSource],
    });

    container.env.copyFrom(secretSource);

    const spec: k8s.Container = container._toKube();

    expect(spec.envFrom).toEqual([
      { configMapRef: { name: cm.name }, prefix: 'pref' },
      { secretRef: { name: secret.name } },
    ]);

    expect(container.env.sources).toEqual([cmSource, secretSource]);
  });

  test('can add environment variables from a secret', () => {});

  test('Can mount container to volume', () => {

    const container = new kplus.Container({
      image: 'image',
    });

    const volume = kplus.Volume.fromConfigMap(kplus.ConfigMap.fromConfigMapName('ConfigMap'));

    container.mount('/path/to/mount', volume);

    const expected: k8s.VolumeMount = {
      mountPath: '/path/to/mount',
      name: volume.name,
    };

    expect(container._toKube().volumeMounts).toEqual([expected]);
  });

  test('can mount container to a pv', () => {

    const chart = Testing.chart();
    const pod = new kplus.Pod(chart, 'Pod');

    const volume = new kplus.AwsElasticBlockStorePersistentVolume(chart, 'PV', { volumeId: 'vol' });

    const container = pod.addContainer({ image: 'image' });
    container.mount('/path/to/mount', volume);

    const resources = Testing.synth(chart);
    expect(resources).toMatchSnapshot();
  });

  test('mount options', () => {
    const container = new kplus.Container({
      image: 'image',
    });

    const volume = kplus.Volume.fromConfigMap(kplus.ConfigMap.fromConfigMapName('ConfigMap'));

    container.mount('/path/to/mount', volume, {
      propagation: kplus.MountPropagation.BIDIRECTIONAL,
      readOnly: true,
    });

    const expected: k8s.VolumeMount = {
      mountPath: '/path/to/mount',
      name: volume.name,
      mountPropagation: 'Bidirectional',
      readOnly: true,
    };

    expect(container._toKube().volumeMounts).toEqual([expected]);
  });

  test('mount from ctor', () => {
    const container = new kplus.Container({
      image: 'image',
      volumeMounts: [
        {
          path: '/foo',
          volume: kplus.Volume.fromEmptyDir('empty'),
          subPath: 'subPath',
        },
      ],
    });

    const expected: k8s.VolumeMount = {
      mountPath: '/foo',
      name: 'empty',
      subPath: 'subPath',
    };

    expect(container._toKube().volumeMounts).toEqual([expected]);
  });

  test('"readiness", "liveness", and "startup" can be used to define probes', () => {
    // GIVEN
    const container = new kplus.Container({
      image: 'foo',
      readiness: kplus.Probe.fromHttpGet('/ping', {
        timeoutSeconds: cdk8s.Duration.minutes(2),
      }),
      liveness: kplus.Probe.fromHttpGet('/live', {
        timeoutSeconds: cdk8s.Duration.minutes(3),
      }),
      startup: kplus.Probe.fromHttpGet('/startup', {
        timeoutSeconds: cdk8s.Duration.minutes(4),
      }),
    });

    // THEN
    expect(container._toKube().readinessProbe).toEqual({
      failureThreshold: 3,
      httpGet: { path: '/ping', port: k8s.IntOrString.fromNumber(80) },
      initialDelaySeconds: undefined,
      periodSeconds: undefined,
      successThreshold: undefined,
      timeoutSeconds: 120,
    });
    expect(container._toKube().livenessProbe).toEqual({
      failureThreshold: 3,
      httpGet: { path: '/live', port: k8s.IntOrString.fromNumber(80) },
      initialDelaySeconds: undefined,
      periodSeconds: undefined,
      successThreshold: undefined,
      timeoutSeconds: 180,
    });
    expect(container._toKube().startupProbe).toEqual({
      failureThreshold: 3,
      httpGet: { path: '/startup', port: k8s.IntOrString.fromNumber(80) },
      initialDelaySeconds: undefined,
      periodSeconds: undefined,
      successThreshold: undefined,
      timeoutSeconds: 240,
    });
  });

  test('Can add resource limits and requests', () => {
    const container = new kplus.Container({
      resources: {
        cpu: {
          request: Cpu.millis(300),
          limit: Cpu.units(0.5),
        },
        memory: {
          request: Size.mebibytes(256),
          limit: Size.mebibytes(384),
        },
      },
      image: 'image',
    });

    expect(container._toKube().resources).toEqual({
      limits: {
        cpu: k8s.Quantity.fromString('0.5'),
        memory: k8s.Quantity.fromString('384Mi'),
      },
      requests: {
        cpu: k8s.Quantity.fromString('300m'),
        memory: k8s.Quantity.fromString('256Mi'),
      },
    });

  });

});

test('Can add only resource requests', () => {
  const container = new kplus.Container({
    resources: {
      cpu: {
        request: Cpu.millis(300),
      },
      memory: {
        request: Size.mebibytes(128),
      },
    },
    image: 'image',
  });

  expect(container._toKube().resources).toEqual({
    requests: {
      cpu: k8s.Quantity.fromString('300m'),
      memory: k8s.Quantity.fromString('128Mi'),
    },
  });
});

test('Can add only resource limits', () => {
  const container = new kplus.Container({
    resources: {
      cpu: {
        limit: Cpu.millis(500),
      },
      memory: {
        limit: Size.mebibytes(1024),
      },
    },
    image: 'image',
  });

  expect(container._toKube().resources).toEqual({
    limits: {
      cpu: k8s.Quantity.fromString('500m'),
      memory: k8s.Quantity.fromString('1024Mi'),
    },
  });
});

test('Can add only limits and requests on memory', () => {
  const container = new kplus.Container({
    resources: {
      memory: {
        limit: Size.mebibytes(1024),
        request: Size.mebibytes(512),
      },
    },
    image: 'image',
  });

  expect(container._toKube().resources).toEqual({
    limits: {
      memory: k8s.Quantity.fromString('1024Mi'),
    },
    requests: {
      memory: k8s.Quantity.fromString('512Mi'),
    },
  });
});

test('Can add only limits and requests on cpu', () => {
  const container = new kplus.Container({
    resources: {
      cpu: {
        limit: Cpu.units(1),
        request: Cpu.millis(250),
      },
    },
    image: 'image',
  });

  expect(container._toKube().resources).toEqual({
    limits: {
      cpu: k8s.Quantity.fromString('1'),
    },
    requests: {
      cpu: k8s.Quantity.fromString('250m'),
    },
  });
});

test('default security context', () => {

  const container = new Container({ image: 'image' });

  expect(container.securityContext.ensureNonRoot).toBeFalsy();
  expect(container.securityContext.privileged).toBeFalsy();
  expect(container.securityContext.readOnlyRootFilesystem).toBeFalsy();
  expect(container.securityContext.user).toBeUndefined();
  expect(container.securityContext.group).toBeUndefined();

  expect(container._toKube().securityContext).toEqual(container.securityContext._toKube());
  expect(container.securityContext._toKube()).toStrictEqual({
    privileged: container.securityContext.privileged,
    readOnlyRootFilesystem: container.securityContext.readOnlyRootFilesystem,
    runAsGroup: container.securityContext.group,
    runAsNonRoot: container.securityContext.ensureNonRoot,
    runAsUser: container.securityContext.user,
  });

});

test('custom security context', () => {

  const container = new Container({
    image: 'image',
    securityContext: {
      ensureNonRoot: true,
      readOnlyRootFilesystem: true,
      privileged: true,
      user: 1000,
      group: 2000,
    },
  });

  expect(container.securityContext.ensureNonRoot).toBeTruthy();
  expect(container.securityContext.privileged).toBeTruthy();
  expect(container.securityContext.readOnlyRootFilesystem).toBeTruthy();
  expect(container.securityContext.user).toEqual(1000);
  expect(container.securityContext.group).toEqual(2000);

});

test('can configure a postStart lifecycle hook', () => {

  const container = new Container({
    image: 'image',
    lifecycle: {
      postStart: Handler.fromCommand(['hello']),
    },
  });

  const spec = container._toKube();
  expect(spec.lifecycle!.postStart).toEqual({ exec: { command: ['hello'] } });

});

test('can configure a preStop lifecycle hook', () => {

  const container = new Container({
    image: 'image',
    lifecycle: {
      preStop: Handler.fromCommand(['hello']),
    },
  });

  const spec = container._toKube();
  expect(spec.lifecycle!.preStop).toEqual({ exec: { command: ['hello'] } });

});

