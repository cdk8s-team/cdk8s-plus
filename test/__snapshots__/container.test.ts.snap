// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container can mount container to a pv 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "kind": "Pod",
    "metadata": Object {
      "labels": Object {
        "cdk8s.io/metadata.addr": "test-Pod-c815bc91",
      },
      "name": "test-pod-c890e1b8",
    },
    "spec": Object {
      "automountServiceAccountToken": true,
      "containers": Array [
        Object {
          "image": "image",
          "imagePullPolicy": "Always",
          "name": "main",
          "resources": Object {
            "limits": Object {
              "cpu": "1500m",
              "memory": "2048Mi",
            },
            "requests": Object {
              "cpu": "1000m",
              "memory": "512Mi",
            },
          },
          "securityContext": Object {
            "privileged": false,
            "readOnlyRootFilesystem": false,
            "runAsGroup": 26000,
            "runAsNonRoot": false,
            "runAsUser": 25000,
          },
          "volumeMounts": Array [
            Object {
              "mountPath": "/path/to/mount",
              "name": "pvc-pvc-test-pv-c8b2a2c6",
            },
          ],
        },
      ],
      "dnsPolicy": "ClusterFirst",
      "securityContext": Object {
        "fsGroupChangePolicy": "Always",
        "runAsNonRoot": false,
      },
      "setHostnameAsFQDN": false,
      "volumes": Array [
        Object {
          "name": "pvc-pvc-test-pv-c8b2a2c6",
          "persistentVolumeClaim": Object {
            "claimName": "pvc-test-pv-c8b2a2c6",
            "readOnly": false,
          },
        },
      ],
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "PersistentVolume",
    "metadata": Object {
      "name": "test-pv-c8b2a2c6",
    },
    "spec": Object {
      "awsElasticBlockStore": Object {
        "fsType": "ext4",
        "readOnly": false,
        "volumeID": "vol",
      },
      "claimRef": Object {
        "name": "pvc-test-pv-c8b2a2c6",
      },
      "persistentVolumeReclaimPolicy": "Retain",
      "volumeMode": "Filesystem",
    },
  },
  Object {
    "apiVersion": "v1",
    "kind": "PersistentVolumeClaim",
    "metadata": Object {
      "name": "pvc-test-pv-c8b2a2c6",
    },
    "spec": Object {
      "volumeMode": "Filesystem",
      "volumeName": "test-pv-c8b2a2c6",
    },
  },
]
`;
